Fixes #30820 - drop digest and use hash index instead

lzap opened the pull request:
Foreman reports storage is designed so that reports are broken into individual lines and store them as 1:N references. The biggest offender are messages and sources tables which contain those lines together with SHA1 hash. This was designed around Puppet because it returns "source" for each log line (on terminal it is printed as "[source] log_content"). So what we essentially do is break reports into lines, break lines into two pieces, calculate SHA1 sum from both parts and store that into the database. Extremely inefficient.

The reason for all of that is that users can search for a source or for a (whole) report line using scoped search. This is only useful for Puppet users (source), I am unable to find a use case for searching for whole line. Full text search would make sense, or doing SQL LIKE query as well, but non of that Foreman supports.

The way this is implemented is to overcome SQL server limit of index because "messages"."value" and "sources"."value" fields are TEXT type. Most SQL servers, including PostgreSQL, can create B-Tree index only up to some limit (8k in this case). This is extremely slow, it's wasting of resources, there is no resolution of SHA1 conflicts.

This patch gets rid of our own SHA1 "digest" column and let SQL server do the index directly on the text column. To break the index limit, Postgres has a hash index option which is not only faster and way more storage-efficient but it's ideal for comparison only, which is exactly what we need (we do not sort logs or sources tables - it makes no sense - we only do equal operation to find duplicities).

The patch

* drops digest completely from the code base and let SQL server to do hashing via index
* drops digest field and index from both tables (2 indices, 2 fields)
* creates a hash index on value column (migration can be slow - testing needed)

The result needs to be tested on real data, but the patch essentially removes one field, removes SHA1 slow calculation in Ruby and replaces two b-tree indices with two hash indices which takes approximately 20-30% of the original space and are faster.

The only issue could be sqlite3 which we still require for packaging, from my rough testing I am able to create index on text column just fine. Initially I thought that creating hash index in RoR migration would require SQL, but it looks like RoR creates the index with the correct type just fine with "using" attribute.

For the record, hash index is supported in postgres from version 8.2 and from version 10 it is transaction safe. Meaning that a crash on pre-10 version would require index rebuild which is a trivial operation. CentOS installations already use Postgres v12 and Debian (Buster) has v11 and Ubuntu has v10. We are all good here.

Also, complete refactoring of reports is planned where searching in reports will be only possible via tablescan anyway:

https://community.theforeman.org/t/rfc-optimized-reports-storage/15573

Finally, the same change must be done in OpenSCAP where digest is being used.

lzap committed:
Fixes #30820 - drop digest and use hash index instead

lzap commented on app/models/report.rb:
```diff
@@ -25,7 +25,6 @@ def self.inherited(child)
       scoped_search :relation => :organization, :on => :name, :complete_value => true, :rename => :organization
       scoped_search :relation => :location,    :on => :name,  :complete_value => true, :rename => :location
-      scoped_search :relation => :messages,    :on => :value,                          :rename => :log, :only_explicit => true
       scoped_search :relation => :sources,     :on => :value,                          :rename => :resource, :only_explicit => true
```
As I said in the description, I don't find searching for exact log lines useful, I don't know why it is here. Searching for sources is useful, the index is still there.

ares replied:
perhaps searching for file name that has changed, however I never used this, does dropping this actually cause any performance gain? or it would be just terribly slow now?

ekohl replied:
When talking to users about this, there were people who explicitly search for things in the logs. I know they told me they searched for exact messages. For example, something failed on one server and they wonder if it also failed on another server. I'm not sure about the source.

tbrisker replied:
This search is not for exact log lines, you can search for partial matches as well, e.g. `log ~ error`

lzap replied:
I will keep this line then, SQL LIKE is still useful. It is explicit only so no big deal.

lzap commented on app/models/source.rb:
```diff
@@ -1,15 +1,17 @@
+
+  def skip_strip_attrs
+    ['value']
   end
```
Skip attribute stripping too to match message model, also saves few CPU cycles during processing as there are many of these usually.

lzap commented on db/migrate/20200911083030_use_hash_index_for_reports.rb:
```diff
@@ -0,0 +1,11 @@
+    add_index(:messages, :value, using: 'hash')
+    add_index(:sources, :value, using: 'hash')
+  end
+end
```
Adding index can be lengthy operation on big instances, we should probably implement a maintain check or something like that asking users to drop old reports prior upgrading.

ekohl commented:
> The only issue could be sqlite3 which we still require for packaging

This is no longer true. Since Foreman 2.1 we use [nulldb](https://github.com/nulldb/nulldb) in packaging and dropped sqlite3 completely. Even less of a concern.

> For the record, hash index is supported in postgres from version 8.2 and from version 10 it is transaction safe. Meaning that a crash on pre-10 version would require index rebuild which is a trivial operation. CentOS installations already use Postgres v12 and Debian (Buster) has v11 and Ubuntu has v10. We are all good here.

https://theforeman.org/manuals/2.1/index.html#3.1SystemRequirements also states "PostgreSQL version 10 or newer" so there can be no doubt about this requirement.

lzap commented:
I have fixed migration rollback (it was missing a type so rollback was not possible) and added an arbitrary benchmark. It does not show off much of an improvement with such a small dataset tho, at least there is not a regression.

```
Warming up --------------------------------------
       import report    28.000  i/100ms
Calculating -------------------------------------
       import report    275.926  (± 5.8%) i/s -      8.260k in  30.043722s
```

With the patch:

```
Warming up --------------------------------------
       import report    28.000  i/100ms
Calculating -------------------------------------
       import report    279.144  (± 4.7%) i/s -      8.372k in  30.056714s
```

This is test environment postgres running without fsync optimized for speed rather than reliability.

lzap commented:
This really needs production tests on a large dataset. The patch does not appear to have any effect on (almost) empty database. I have rewritten the patch to be integration (end-to-end) test. Before the patch:

```
[lzap@box foreman]$ ruby test/benchmark/puppet_report_benchmark.rb
Calculating -------------------------------------
import report via HTTP
                          0.349  (± 0.0%) i/s -     21.000  in  60.759114s
```

After the patch:

```
[lzap@box foreman]$ ruby test/benchmark/puppet_report_benchmark.rb
Calculating -------------------------------------
import report via HTTP
                          0.351  (± 0.0%) i/s -     21.000  in  60.032783s
```

The report import performance is still abysmal, 20 simple reports per minute. However with hash indices amount of data stored in database should be smaller.

lzap commented:
Okay, I was digging in the codebase for the whole Friday and it looks like we can improve quite a bit here. Here is the base performance with the hash index (which does not have a huge impact on small scale installations):

```
[lzap@box foreman]$ ruby test/benchmark/puppet_report_benchmark.rb
Calculating -------------------------------------
import report via HTTP
                          0.268  (± 0.0%) i/s -      6.000  in  22.441481s
```

First, I have identified that report create endpoint renders the input back via RABL and it's extremely slow. A small patch in the controller which returns nothing (empty hash) instead of the processed report almost doubles the performance. This small change can easily be backported into old versions, however it breaks the API which is okay for our ENC script but users might be parsing the output. Workaround is to fetch the inserted report in an extra call:

```
Calculating -------------------------------------
import report via HTTP
                          0.410  (± 0.0%) i/s -      9.000  in  22.017360s
```

On top of that I have another patch, using Rails 6 new feature called "batch inserts" I am able to improve the performance 5x. The drawback is that the processing code is much more complex now:

```
Calculating -------------------------------------
import report via HTTP
                          1.491  (± 0.0%) i/s -     30.000  in  20.231874s
```

lzap commented on app/controllers/api/v2/config_reports_controller.rb:
```diff
@@ -40,7 +40,10 @@ def show
+        errors = @config_report.errors.empty?
+        # RABL is extremely slow for huge reports and ENC script does not care about the output
+        @config_report = {}
+        process_response errors
```
These few lines almost doubles performance, drawback: the endpoint now returns nothing instead of the processed report. Another call needed to fetch the report from DB.

We should look into why RABL is so slow as well.

tbrisker replied:
I think the reason is that the `show.rabl` which is returned here relies on all the child attributes which are pulled inside the view processing likely triggering an n+1 query:
https://github.com/theforeman/foreman/blob/develop/app/views/api/v2/config_reports/show.json.rabl#L5-L13

ekohl commented on app/controllers/api/v2/config_reports_controller.rb:
```diff
@@ -40,7 +40,10 @@ def show
         @config_report = ConfigReport.import(params.to_unsafe_h[:config_report], detected_proxy.try(:id))
-        process_response @config_report.errors.empty?
+        errors = @config_report.errors.empty?
+        # RABL is extremely slow for huge reports and ENC script does not care about the output
```
This is confusing. The ENC script has nothing to do with config reports.

Also note that the config reports controller is not just for Puppet reports. It's also used by Ansible and possibly others. It's breaking the API and as such it must at least be accompanied by an upgrade warning.

I think this may deserve its own issue so it properly shows up in the changelog.

lzap replied:
I state above that the change is breaking the API and I know that other plugins use this endpoint. True this should be split up from this PR. Luckily it is completely isolated and does not depend on hash-index change. Pulling it into a separate PR (note for me sha `62cbab1`)

lzap replied:
Removing from there, the PR is: https://github.com/theforeman/foreman/pull/7987

ekohl commented on app/services/report_importer.rb:
```diff
@@ -91,21 +91,60 @@ def logs
+      msg = log['log']['messages']['message']
+      src = log['log']['sources']['source']
+
+      msg_id = Message.where(value: msg).pluck(:id).first
```
This does a ton of database queries which is typically very slow, especially on remote databases. It's probably faster to do a single DB query for all messages and see which already exist. Same for Source below.

lzap replied:
Yes I was thinking the same, but you can have thousands of facts (we've seen 10,000 facts for a host with 1400 HDDs) and this will definitely break some limit of SQL query. I'd probably need to break this into some kind of batches, this will make the code even more complex. Do we really want to do that?

On the other hand, we can set some maximum mount of facts. Not sure how a host with 10,000 facts is actually useful. Postgres also has quite reasonable length limits and it appears to be configurable. Let me think about this and try some more testing if it helps to improve performance even further.

ekohl replied:
We shouldn't mix up facts and reports. This is just reports and reports have no facts. Messages shouldn't be dropped.

A report can have 100s or 1000s of messages. In that case, we will do 1000s of SQL queries and that's known to be slow.

However, we also know that a huge `WHERE IN ( .... )` isn't great. You can consider temporary tables and join.

lzap replied:
Sorry disregard my comment about facts, I find hard to switch from facts to reports. I often work on these areas :-)

Anyway, a report can have many lines therefore we would generate a huge SQL statement. I am not sure about temporary tables if it is worth the effort. I need to take a look.

lzap replied:
I was looking into possibility of a single query and I cannot do it. The idea behind this optimization is to find out if an existing message/source is already present. I cannot do this in a single call because SQL servers randomize output when you use "WHERE ... IN" statement. There are some really complex solutions (using special sorted arrays) for postgres, but I don't think this is worth the effort.

Another problem is that the SQL query contains whole message/source line as the hash is now calculated in the SQL driver/server. Therefore this can make the single query very long, we are not transferring just IDs but long lines.

The patch still shows good improvement, we used to do ton of queries and then ton of insert statements. Now it leverages `insert_all` Rails method which performs inserting in batches. This is very important specifically for the join table (logs).

ekohl commented:
> This small change can easily be backported into old versions, however it breaks the API which is okay for our ENC script but users might be parsing the output.

As indicated - it's breaking the API and IMHO certainly can't be backported into a stable release. The ENC script doesn't care because it's completely unrelated to reports. The Puppet report processor doesn't care either, but other consumers might. I have no idea what Ansible and Chef do for example, but pretty sure they also use this endpoint.

If you're not going to return anything, it may be better to at least add a header `Location: /path/to/report/with/id` so API consumers at least know which report ID was created.

As suggested inline: this deserves its own review. It doesn't look like these 2 matters need to be in the same PR.

tbrisker commented on app/models/message.rb:
```diff
@@ -2,15 +2,13 @@ class Message < ApplicationRecord
-    digest = Digest::SHA1.hexdigest(val)
-    Message.find_by(:digest => digest) || Message.create(:value => val, :digest => digest)
+  def self.find_or_create(value)
+    Message.find_or_create_by(value: value)
```
nit: the `Message` is not needed
```suggestion
    find_or_create_by(value: value)
```
Should we deprecate this method and use `find_or_create_by` directly?

lzap replied:
Breaks at least two plugins, so deprecation message + this. But problem is this will flood logs with a lot of warnings :-(

tbrisker commented on app/models/source.rb:
```diff
@@ -1,15 +1,17 @@
-    digest = Digest::SHA1.hexdigest(val)
-    Source.find_by(:digest => digest) || Source.create(:value => val, :digest => digest)
+  def self.find_or_create(value)
+    Source.find_or_create_by(value: value)
```
here as well

lzap replied:
Fixed both, adding just a simple deprecation comment until we fix plugins.

tbrisker commented on app/services/report_importer.rb:
```diff
@@ -91,21 +91,60 @@ def logs
   end
 
   def import_log_messages
+    id_messages = []
```
I think using hashes here with a reasonable key (i.e. message or source value) instead of arrays can both make the lookups faster and the code easier to understand.

lzap replied:
I don't actually search in arrays, I just append once and iterate once over them. If you have a better idea let me know how to do this.

tbrisker reviewed (changes requested):
For the record, if i'm not mistaken the reason we added hashes was that mysql couldn't index text fields over a certain length, so finding matching messages would trigger a full table scan instead of index check. Considering they aren't sorted using a hash index instead of btree definitely makes sense.

lzap commented:
> it's breaking the API and IMHO certainly can't be backported into a stable release

Agreed, I am not proposing to backport anything. It's just we have a customer struggling with performance and we might do a custom hotfix for them - that's a different discussion, sorry for the confusion.

> If you're not going to return anything, it may be better to at least add a header `Location: /path/to/report/with/id` so API consumers at least know which report ID was created.

> If you're not going to return anything, it may be better to at least add a header Location: /path/to/report/with/id so API consumers at least know which report ID was created.

Good idea, will do in the other PR.

lzap commented:
TODO:

* [x] fix unit tests
* [x] try to do single query pluck and test its performance
* [x] drop unique Rails validation + index + test from reported_at column (?)

lzap commented on app/models/report.rb:
```diff
@@ -17,15 +17,14 @@ class Report < ApplicationRecord
 
   validates :host_id, :status, :presence => true
-  validates :reported_at, :presence => true, :uniqueness => {:scope => [:host_id, :type]}
+  validates :reported_at, :presence => true
```
Do you understand why this was here? This caused one test to fail:

```ruby
    def test_create_duplicate
      User.current = nil
      post :create, params: { :config_report => create_a_puppet_transaction_report }, session: set_session_user
      assert_response :success
      post :create, params: { :config_report => create_a_puppet_transaction_report }, session: set_session_user
      assert_response :unprocessable_entity
    end
```

What is the point here? I don't see any reason why the application should block uploading the same report twice, it only requires the server to search through the index. I suggest to drop this uniquness test + the index as well + the test.

lzap replied:
This is probably not a performance problem so I am keeping this. Let's remove this in a separate PR if needed.

ares replied:
I assume the validation is there to avoid duplicate report issues. I know there were other places in the code preventing uploads too often. Then we needed to add this, because we had more types. This validation prevents ArfReport and ConfigReport creation if they were created at the same time (it happened). It's still not bulletproof, since Ansible config report can in theory appear at the same time as Puppet config report. The whole validation IMHO is useless.

lzap commented:
This is now ready for re-review.

lzap commented:
I have created PL/pgSQL procedure which creates reports with random data. My testing shows that a simple change of indices from btree to hash boosts the performance of report creation quite a bit. As the data grows, the difference is bigger and bigger. I was able to create 3k reports with 500 messages and 500 sources (all unique) in 6 minutes (btree index) and in 1m42s (hash index). That's a huge.

The whole analysis is here:

https://gist.github.com/lzap/0fb83aa93728cbb67d4b11345da7c1bd

It is also worth noting that users who are suffering from slow reports can easily change indices from btree to hash on their existing installations, this will not break the migration in the future once this patch is merged (index will be dropped once again). Here are the statements which needs to be done:

```
su postgres -c "psql foreman"
foreman=# drop index index_messages_on_digest;
DROP INDEX
foreman=# create index index_messages_on_digest on messages using hash (digest);
CREATE INDEX
foreman=# drop index index_sources_on_digest;
DROP INDEX
foreman=# create index index_sources_on_digest on sources using hash (digest);
CREATE INDEX
```

According to my rough measurements on my VM, hash index creation was as fast as 18 million records per minute. I measured 10 seconds to index 3 millions and since hash is O(n) it should be pretty linear. This is good news for our migration - it can be pretty fast.

lzap commented:
Tests are green, I would like to keep the other optimization as a separate commit (it depends on the other one tho).

lzap commented:
Anyone? I would not suggest merging this days before branching as I need to followup with some upgrade instructions etc. But according to my testing upgrade is pretty fast even for larger datasets.

lzap commented:
Hello @tbrisker with branching incoming is there any chance getting this (and the other related PR) in? I am pretty sure I will not be able to deliver report refactoring for 2.3 and this should be a nice improvement for puppet-heavy deployments.

tbrisker reviewed (changes requested):
@lzap can you split off the hash instead of digest to a separate pr? that part can go in to 2.3, the import optimization is a larger change I'd prefer giving more time to mature in nightlies to make sure it's working well.

lzap commented:
Dropped the optimization commit.

lzap commented:
https://github.com/theforeman/foreman/pull/8102

tbrisker commented on app/models/message.rb:
```diff
@@ -2,15 +2,14 @@ class Message < ApplicationRecord
   has_many :reports, :through => :logs
   has_many :logs
   validates_lengths_from_database
-  validates :value, :digest, :presence => true
```
is dropping the `value` validation on purpose?

lzap replied:
Yes, I recall I was having some issues with tests. Let me remove this to see if they are passing.

lzap replied:
It had something to do with the database index, presense validator requires an unique index and hash index indeed is exactly (only) unique.

lzap commented on app/models/source.rb:
```diff
@@ -1,15 +1,19 @@
   has_many :reports, :through => :logs
   has_many :logs
+  validates_lengths_from_database
   validates :value, :digest, :presence => true
```
This matches the Message class so they are exactly the same.

tbrisker replied:
they should both be:
```suggestion
  validates :value, :presence => true
```
I believe, we still want to make sure the value is there but the digest is gone

lzap replied:
Damn right thanks.

lzap commented:
Rebased, let's see. Again, I tested the migration on a decent dataset and it seems to be quick enough so no worries there.

tbrisker reviewed (approved):
Thanks @lzap !

lzap commented:
I will do an upgrade note too.
